### PR TITLE
[9.0][MIG] medical prescription sale

### PR DESCRIPTION
**Pull request by lasley**
This is dependent on #79 and #77

This module adds a wizard to create quotations from prescription lines. 

Note that I had to add `wkhtmltopdf` to the apt dependencies in Travis in order to green the tests for sale module

Depends:
- [x] #77 
- [x] #79

### CODE REVIEW

File: medical/models/medical_patient.py
Comment (JonathanNEMRY) [inline medical/models/medical_patient.py]: remove comma

File: medical/models/medical_patient.py
Comment (JonathanNEMRY) [inline medical/models/medical_patient.py]: remove comma

File: medical_pathology/__init__.py
Comment (JonathanNEMRY) [inline medical_pathology/__init__.py]: Use the correct OCA header like https://github.com/OCA/maintainer-tools/blob/master/template/module/exceptions.py#L2

File: medical_pathology/tests/__init__.py
Comment (JonathanNEMRY) [inline medical_pathology/tests/__init__.py]: same comment for the header

Comment (lasley) [inline medical_pathology/tests/__init__.py]: Thanks @JonathanNEMRY - Will apply this to all PRs, and rebase accordingly

**Comment by lasley** [comment]
Note this migration should also include laslabs#51

**Comment by woodbrettm** [comment]
This is ready, minus lint check fails and rebase after prescription thread is merged

**Comment by lasley** [comment]
Rebased now that #141 is merged

**Comment by woodbrettm** [comment]
lol dang was just doing that xD. Will fix the lint errors.

File: sale_medical_prescription/wizards/medical_sale_wizard.py
Comment (woodbrettm) [inline sale_medical_prescription/wizards/medical_sale_wizard.py]: Remove these comments, as well as those on 89 and 90. Already in readme.

File: sale_medical_prescription/wizards/medical_sale_wizard.py
Comment (woodbrettm) [inline sale_medical_prescription/wizards/medical_sale_wizard.py]: @lasley these were changed from `patient_id.id` to `partner_id.id`. During testing I had initially tested against patients, however the tests showed the results were pointing to related partner of patient.

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (woodbrettm) [inline sale_medical_prescription/wizards/medical_sale_temp.py]: Remove comments. @lasley do we need `payment_term` and `fiscal_position`? If so, should I put it in the roadmap?

File: sale_medical_prescription/README.rst
Comment (woodbrettm) [inline sale_medical_prescription/README.rst:14]: @lasley the readme has not been updated yet to reflect your opened issue #147

File: sale_medical_prescription/models/medical_medicament.py
Comment (woodbrettm) [inline sale_medical_prescription/models/medical_medicament.py]: `categ_id` is a required field AFAIK so removed this check. @lasley pointing out in case you want me to keep it.

**Review by woodbrettm — COMMENTED** [review]
This points out a few notable changes. Thing or two needs to be fixed.

**Comment by woodbrettm** [comment]
ah woops looks like I may have commented on my first commit hence the collapsed comments. All are still relevant, however. Ready for review!

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (woodbrettm) [reply sale_medical_prescription/wizards/medical_sale_temp.py]: was getting warnings that those fields don't exist in sale.order

File: sale_medical_prescription/__openerp__.py
Comment (woodbrettm) [inline sale_medical_prescription/__openerp__.py]: As per oca template, wizard views should be moved back to wizards folder

File: sale_medical_prescription/demo/medical_medicament_demo.xml
Comment (woodbrettm) [inline sale_medical_prescription/demo/medical_medicament_demo.xml:13]: medicaments should all be product

Comment (lasley) [reply sale_medical_prescription/demo/medical_medicament_demo.xml:13]: This is incorrect; product is a stock dependency. Leave as is

File: sale_medical_prescription/README.rst
Comment (lasley) [reply sale_medical_prescription/README.rst:14]: In this case, we're actually fine. The only concept added is documented

File: sale_medical_prescription/models/medical_patient.py
Comment (lasley) [inline sale_medical_prescription/models/medical_patient.py]: Move this to roadmap

File: sale_medical_prescription/models/medical_pharmacy.py
Comment (lasley) [inline sale_medical_prescription/models/medical_pharmacy.py]: `_compute_verified_by_id_and_date`

File: sale_medical_prescription/models/medical_physician.py
Comment (lasley) [inline sale_medical_prescription/models/medical_physician.py]: _compute_verified_by_id_and_date

File: sale_medical_prescription/models/medical_physician.py
Comment (lasley) [inline sale_medical_prescription/models/medical_physician.py:11]: Add to readme:
`* Move duplicated `is_verified`, `verified_by_id`, `verified_date` & related methods to `medical.entity.abstract`

File: sale_medical_prescription/models/medical_pharmacy.py
Comment (lasley) [inline sale_medical_prescription/models/medical_pharmacy.py]: Do a string replace in this module to switch all the `rec_id` to `record`

File: sale_medical_prescription/models/sale_order.py
Comment (lasley) [inline sale_medical_prescription/models/sale_order.py]: `prescriptions = rx_model.browse()` (same with lines)

File: sale_medical_prescription/models/sale_order.py
Comment (lasley) [inline sale_medical_prescription/models/sale_order.py]: `prescription_lines += line`

File: sale_medical_prescription/models/sale_order.py
Comment (lasley) [inline sale_medical_prescription/models/sale_order.py]: `rec_id.prescription_order_ids = prescriptions`

File: sale_medical_prescription/wizards/medical_sale_line_temp.py
Comment (lasley) [inline sale_medical_prescription/wizards/medical_sale_line_temp.py]: Remove this line

File: sale_medical_prescription/wizards/medical_sale_wizard.py
Comment (lasley) [inline sale_medical_prescription/wizards/medical_sale_wizard.py]: Move to roadmap

**Review by lasley — COMMENTED** [review]
Thanks @bmw95 - comments inline

@t3ddftw please review

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (tedsalmon) [inline sale_medical_prescription/wizards/medical_sale_temp.py]: Imports should be sorted alphabetically.

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (tedsalmon) [inline sale_medical_prescription/wizards/medical_sale_temp.py]: Remove commented line. s/rec_id/record

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (tedsalmon) [inline sale_medical_prescription/wizards/medical_sale_temp.py]: Remove commented line

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (tedsalmon) [inline sale_medical_prescription/wizards/medical_sale_temp.py]: Remove commented line

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (tedsalmon) [inline sale_medical_prescription/wizards/medical_sale_temp.py]: Move to README

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (tedsalmon) [inline sale_medical_prescription/wizards/medical_sale_temp.py]: Remove commented code. Move relevant `@TODO` to README

**Review by tedsalmon — CHANGES_REQUESTED** [review]
Minor changes requested.

File: sale_medical_prescription/__openerp__.py
Comment (lasley) [inline sale_medical_prescription/__openerp__.py]: 9.0.1.0.0

File: sale_medical_prescription/demo/sale_order_demo.xml
Comment (woodbrettm) [inline sale_medical_prescription/demo/sale_order_demo.xml]: [Travis Build Fail](https://travis-ci.org/OCA/vertical-medical/jobs/187223678#L1196)
`"External ID not found in the system: portal.demo_user0_res_partner" while parsing /home/travis/build/OCA/vertical-medical/sale_medical_prescription/demo/sale_order_demo.xml:7`

File: sale_medical_prescription/models/sale_order.py
Comment (woodbrettm) [reply sale_medical_prescription/models/sale_order.py]: @lasley should I do `rec_id.prescription_order_ids = set(prescriptions)`? Wondering if there's a possibility of duplicates.

File: sale_medical_prescription/wizards/medical_sale_temp.py
Comment (woodbrettm) [reply sale_medical_prescription/wizards/medical_sale_temp.py]: Nice catch, thought I got all of them lol.

**Comment by woodbrettm** [comment]
## Updated.
I'm not sure if this is an issue or not but discovered if I add multiple sale order lines in the wizard and click create I get this error. Seems that uom is not being carried over? Not sure.
* Orders are divided by patient
* Issue was replicated in laslabs/release/9.0
* Fixed for now at least, by disabling adding any more items (pic below is with adding items enabled)
  * If adding items is disabled, the only item can be changed to a different one and no errors happen
  * Other items can be added to the sale order in the sale order view form after redirected there when wizard is done.

sale

**Comment by lasley** [comment]
@bmw95 - are you adding the lines directly into the wizard using "Add an Item", or are you checking the lines in prescription line form view then hitting Create Sale to initiate the wizard with the lines?

**Comment by woodbrettm** [comment]
@lasley in the rx order line form view, I clicked *action* => *Create Quote(s)*, where I was shown the temp sale form below.
hello

Clicked on *Generate Template(s)*, showing me the next page which is the first pic I posted. I then clicked on *Add an item*, which I added the Simvastatin product, then clicked on *Create* thus causing the error.

I tried the process again today, and the logs were showing that the 2nd product, INSERT INTO sale_order_line, null column product_uom.

**Comment by lasley** [comment]
You need to remove the ability for line item creation in the wizard - you should not be able to add some random prescription from there, the process is to select the prescriptions then start the wizard for the ones selected. Use `options="{'no_create': True}"` in the m2m field

**Comment by woodbrettm** [comment]
Okie dokes.

**Comment by woodbrettm** [comment]
Verified cannot add order lines to wizard. Good to go!

File: sale_medical_prescription/models/sale_order.py
Comment (lasley) [inline sale_medical_prescription/models/sale_order.py]: We can make things nicer + fix the coverage drop in one swoop by removing below line and switching this to:
```
for order_line in record.order_line.filtered(lambda r: r.prescription_line_id):
```

Comment (woodbrettm) [reply sale_medical_prescription/models/sale_order.py]: Oh nice. Yeah that's much better.

**Comment by woodbrettm** [comment]
@lasley updated

**Review by lasley — COMMENTED** [review]
LGTM thanks @bmw95 

@tedsalmon can you check that your comments were attended to & approve your review please

**Review by tedsalmon — APPROVED** [review]
Most excellent. 👍

**Pull request merged**
